dandiapi/api/views/version.py

```diff
@@ -102,10 +103,27 @@ def update(self, request, **kwargs):
         responses={200: VersionSerializer()},
     )
     @action(detail=True, methods=['POST'])
+    @transaction.atomic
     @method_decorator(permission_required_or_403('owner', (Dandiset, 'pk', 'dandiset__pk')))
     def publish(self, request, **kwargs):
         """Publish a version."""
-        old_version: Version = self.get_object()
+        # Note: instead of using self.get_object() here, retrieve the Version/perform permission
+        # checks manually. This is so we can use select_for_update to lock rows to prevent
+        # double publishing
+        queryset: QuerySet[Version] = self.get_queryset()
+        try:
+            old_version: Version = get_object_or_404(
+                # `nowait` is used here so that this request aborts and returns a 400 error as soon
+                # as it sees that the target Version is locked. The default behavior of
+                # `select_for_update` is to block until the lock can be acquired, but in this case
+                # we don't need to do that and incur the performance hit that comes with it.
+                queryset.select_for_update(nowait=True),
+                **kwargs,
+            )
+        except DatabaseError:
+            raise ValidationError('Dandiset is already in the process of being published')
+        self.check_object_permissions(self.request, old_version)
```

Comment:
Unfortunately we lost a bit of the convenience/safety that `get_object` provides with automatic permission checking, so it must be called explicitly here. It seems that there's been some upstream discussions about supporting `select_for_update` in `get_object` on both the Django and DRF sides, but nothing has come of it yet. ([Django issue](https://code.djangoproject.com/ticket/27477), [DRF issue](https://github.com/encode/django-rest-framework/issues/4675))

Reply:
The check probably deserves a comment.

Reply:
Added https://github.com/dandi/dandi-archive/pull/1006/commits/f2c41c3fa54edd43db34b043b28ed03d48d62380